docker/templates/devel.template

```diff
@@ -0,0 +1,42 @@
+RUN apt-get update \
+ && DEBIAN_FRONTEND=noninteractive apt-get install -y \
+    libgflags-dev \
+    libgoogle-glog-dev \
+    libopencv-dev \
+    libleveldb-dev \
+    libsnappy-dev \
+    liblmdb-dev \
+    libhdf5-serial-dev \
+    libprotobuf-dev \
+    protobuf-compiler \
+    libatlas-base-dev \
+    python-dev \
+    python-pip \
+    python-numpy \
+    python-scipy \
+    gfortran
+
+# Use --no-install-recommends for the boost libs.
+RUN apt-get install -y --no-install-recommends \
+    libboost-all-dev
+
+# Install utilities used for the build.
+RUN apt-get install -y \
+    build-essential \
+    cmake \
+    cmake-curses-gui \
+    git \
+    ca-certificates \
+    bc \
+    wget
+
+# Install the python requirements.
+RUN wget https://raw.githubusercontent.com/BVLC/caffe/master/python/requirements.txt -O /tmp/requirements.txt && \
```

Comment:
> This means that the caffe source (and thus requirements.txt) is not available to the image at build time.

Why not?  The source is available locally.  How else would you build the image?  Yes, you may need to run `docker build` at a higher level (run it at the root level of the repo, and ignore some things in `.dockerignore`), but everything is visible to the docker build script.

Why not just `ADD` the file to `/tmp/`?  The repository is already checked out.  I see that your `RUN` command avoids a layer in AUFS, but it's a tiny layer containing one file.  Not worth the micro-optimization.

Some reasons why `ADD python/requirements.txt /tmp/` is better than `RUN wget ...`:

1. If editing `requirements.txt` and re-building, the old version will be cached in docker and the new version will not run (because the `RUN` command has not changed).  If you use `ADD` and change `requirements.txt`, re-building the docker image will see the new version of the file.
2. The build will _always_ use the repository version, leading to some unintuitive errors if one tries to edit `requirements.txt` and re-build the image.